include/picongpu/fields/background/templates/twtsfast/BField.tpp

```diff
@@ -0,0 +1,738 @@
+                , w_x_SI(w_x_SI)
+                , phi(phi)
+                , beta_0(beta_0)
+                , tdelay_user_SI(tdelay_user_SI)
+                , dt(SI::DELTA_T_SI)
+                , unit_length(UNIT_LENGTH)
+                , auto_tdelay(auto_tdelay)
+                , pol(pol)
+                , phiPositive(float_X(1.0))
+            {
+                /* Note: Enviroment-objects cannot be instantiated on CUDA GPU device. Since this is done
+                 * on host (see fieldBackground.param), this is no problem.
+                 */
+                SubGrid<simDim> const& subGrid = Environment<simDim>::get().SubGrid();
+                halfSimSize = subGrid.getGlobalDomain().size / 2;
+                tdelay = detail::getInitialTimeDelay_SI(
+                    auto_tdelay,
+                    tdelay_user_SI,
+                    halfSimSize,
+                    pulselength_SI,
+                    focus_y_SI,
+                    phi,
+                    beta_0);
+                if(phi < float_X(0.0))
+                    phiPositive = float_X(-1.0);
+            }
+
+            template<>
+            HDINLINE float3_X BField::getTWTSBfield_Normalized<DIM3>(
+                pmacc::math::Vector<floatD_64, detail::numComponents> const& bFieldPositions_SI,
+                float_64 const time) const
+            {
+                typedef pmacc::math::Vector<float3_64, detail::numComponents> PosVecVec;
+                PosVecVec pos(PosVecVec::create(float3_64::create(0.0)));
+
+                for(uint32_t k = 0; k < detail::numComponents; ++k)
+                {
+                    for(uint32_t i = 0; i < simDim; ++i)
+                    {
+                        pos[k][i] = bFieldPositions_SI[k][i];
+                    }
+                }
+
+                /* An example of intra-cell position offsets is the staggered Yee-grid.
+                 *
+                 * Calculate By-component with the intra-cell offset of a By-field
+                 */
+                float_64 const By_By = calcTWTSBy(pos[1], time);
+                /* Calculate Bz-component the the intra-cell offset of a By-field */
+                float_64 const Bz_By = calcTWTSBz_Ex(pos[1], time);
+                /* Calculate By-component the the intra-cell offset of a Bz-field */
+                float_64 const By_Bz = calcTWTSBy(pos[2], time);
+                /* Calculate Bz-component the the intra-cell offset of a Bz-field */
+                float_64 const Bz_Bz = calcTWTSBz_Ex(pos[2], time);
+                /* Since we rotated all position vectors before calling calcTWTSBy and calcTWTSBz_Ex,
+                 * we need to back-rotate the resulting B-field vector.
+                 *
+                 * RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the field vectors.
+                 */
+                float_64 const By_rot = -math::sin(+phi) * By_By + math::cos(+phi) * Bz_By;
```

Comment:
please use `pmacc::math::sincos()`

Reply:
https://github.com/ComputationalRadiationPhysics/picongpu/blob/f6f15f7e4ff96f2ec3fc7630b7ba021b8ff83a9b/include/pmacc/algorithms/math/defines/trigo.hpp#L36

Reply:
please write `float_X const By_rot ...` to avoid casts in line `123`. This will increase readability.